text/000-rustdoc-jump-to-definition.md

```diff
@@ -0,0 +1,187 @@
+Rustdoc: jump to definition
+
+- Feature Name: `jump_to_definition`
+- Start Date: 2021-12-09
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#89095](https://github.com/rust-lang/rust/issues/89095)
+
+# Summary
+[summary]: #summary
+
+Generate links on idents in the rustdoc source code pages which links to the item's definition and documentation (if available).
+
+# Motivation
+[motivation]: #motivation
+
+This RFC proposes to generate links in the rustdoc source code pages (the ones you end up on when clicking on the `source` links) to allow you to jump to an item definition or back to its documentation page.
+
+The goal of this RFC is to greatly improve the browing experience of the source page (especially when looking for private items).
```

Comment:
This section is still not a motivation.

The motivation is one of the most important parts of the RFC. It should not be written presupposing a conclusion. It is not a place to simply describe the feature. It is a place to, from the ground up, explain _why_ such a feature is desired.

Reply:
It should also be careful explaining who this is for. Is this primarily for users of a crate looking to understand a type better? Or is it primarily for maintainers of the crate? That affects the design of this feature greatly.

For example, if it is primarily for users of a crate; to me it feels like we ought to focus on links in code examples _first_ (but that has its own complicated design problems). If it's primarily for maintainers, that's the scope expansion I'm worried about, rustdoc has historically been a tool for crate _users_ and if we're switching that around we need to be super super explicit about that.

Reply:
cc @jsha who may have some thoughts about picking a target user base

Reply:
> This section is still not a motivation.
>
> The motivation is one of the most important parts of the RFC. It should not be written presupposing a conclusion. It is not a place to simply describe the feature. It is a place to, from the ground up, explain why such a feature is desired.

Ok, I think I understand my misunderstanding. The motivation should be something like "The source code pages usefulness is limited because it doesn't allow to actually browse efficiently through it". Or something along the line. Dissertation is not my forte so I'll need some time to come up with something more readable but that seems like a better start.

> It should also be careful explaining who this is for. Is this primarily for users of a crate looking to understand a type better? Or is it primarily for maintainers of the crate? That affects the design of this feature greatly.
> ...

Interestingly enough, the primary target I have for this feature is users, not crate maintainers, but we arrive at two different conclusions. As a user of a crate, I want to be able to understand how its internals work, and to do so I currently need to clone it locally because I can't easily "follow a lead" with the current source code pages. Which becomes quickly annoying when this same crate uses dependencies, etc.

Side note: I didn't think about adding links in examples. That could be a nice improvement too, but out of scope in the current case (still keeping it in mind for later though).